File: Typescript/src/DelegationSortedIntList.ts
/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 *
 */

import { AbstractIntList } from "./AbstractIntList";
import { IntegerList } from "./IntegerList";
import { SortedIntLinkedList } from "./hidden/SortedIntListLibrary";

class DelegationSortedIntList implements AbstractIntList{
  private l: SortedIntLinkedList
  private totalAdded: number
  constructor() {
    this.l = new SortedIntLinkedList
    this.totalAdded = 0
  }

  add(num: number): boolean {
      if (this.l.add(num)) {
        this.totalAdded += 1
        return true
      }
      return false
  }

  public addAll(list: IntegerList): boolean {
      var old_size = this.l.size()
      var ret = this.l.addAll(list)
      this.totalAdded += (this.l.size() - old_size)
      return ret
  }

  get(index: number): number {
      return this.l.get(index)
  }

  remove(num: number): boolean {
      return this.l.remove(num)
  }

  removeAll(list: IntegerList): boolean {
      return this.l.removeAll(list)
  }

  size(): number {
      return this.l.size()
  }

  getTotalAdded() {
    return this.totalAdded
  }
}

export { DelegationSortedIntList }

File: Java/src/main/java/edu/cmu/cs/cs214/rec04/DelegationSortedIntList.java
package edu.cmu.cs.cs214.rec04;

/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 *
 */

public class DelegationSortedIntList implements IntegerList {
    private SortedIntLinkedList l;
    private int totalAdded;
    
    /**
     * Constructor, create the private var SortedIntLinkedList to store data.
     * Initialize totalAdded as 0.
     */
    DelegationSortedIntList() {
        l = new SortedIntLinkedList();
        totalAdded = 0;
    }

    public boolean add(int num) {
        if (l.add(num)) {
            totalAdded += 1;
            return true;
        }
        return false;
    }

    public boolean addAll(IntegerList list) {
        int old_size = l.size();
        boolean ret = l.addAll(list);
        totalAdded += (l.size() - old_size);
        return ret;
    }

    public int get(int index) {
        return l.get(index);
    }

    public boolean remove(int num) {
        return l.remove(num);
    }

    public boolean removeAll(IntegerList list) {
        return l.removeAll(list);
    }

    public int size() {
        return l.size();
    }

    /**
     * @return The total number of numbers being added.
     */
    public int getTotalAdded() {
        return totalAdded;
    }

}